pymc/initvalues.py

```
@@ -0,0 +1,212 @@
+    ----------
+    initvals : optional, list or dict
+        Initial value strategy overrides that should take precedence over the defaults from the model.
+        Can be None or dict will apply to all chains.
+        A sequence of None or dicts will be treated as chain-wise strategies and must have the same length as `seeds`.
+    rvs_to_jitter : set
+        Random variable tensors for which U(-1, 1) jitter shall be applied.
+        (To the transformed space if applicable.)
+    seeds : list
+        Seeds for initial point evaluation for each chain.
+    """
+    ipfns = None
+    if isinstance(initvals, dict) or initvals is None:
+        # One strategy for all chains
+        # Only one function compilation is needed.
+        ipfns = [
+            model.make_initial_point_fn(
+                jitter_rvs=rvs_to_jitter,
+                return_transformed=True,
+                overrides=convert_start_dict_to_initval_dict(model, initvals or {}),
+            )
+        ] * len(seeds)
+    elif len(initvals) == len(seeds):
+        ipfns = [
+            model.make_initial_point_fn(
+                jitter_rvs=rvs_to_jitter,
+                return_transformed=True,
+                overrides=convert_start_dict_to_initval_dict(model, chain_start or None),
+            )
+            for chain_start in initvals
+        ]
+    else:
+        raise ValueError(
+            f"Number of start dicts ({len(initvals)}) does not match the number of seeds ({len(seeds)})."
+        )
+
+    # The chain-wise seed is respected for the start point evaluation.
+    start_points = [ipfn(seed) for ipfn, seed in zip(ipfns, seeds)]
+    return start_points
+
+
+def make_initial_point_expression(
+    *,
+    free_rvs: Sequence[TensorVariable],
+    rvs_to_values: Dict[TensorVariable, TensorVariable],
+    initval_strategies: Dict[TensorVariable, Optional[Union[np.ndarray, Variable, str]]],
+    jitter_rvs: Set[TensorVariable] = None,
+    default_strategy: str = "prior",
```

Shouldn't the default strategy be moment as in v3?

---

We can't do that until we implemented moments for a ton of RVs. (out of scope of this PR)

---

moment should just default to random/prior if not implemented

---

> moment should just default to random/prior if not implemented

Moment defaults to `NotImplementedError`. Anything else is like asking for trouble, don't you think? I mean if you do `pm.get_moment()` and it gives you a random prior draw that's like... super unexpected and hard to find out what happened

---

Ok then when we call it we can have a try except block. 

I don't think we had moments for all dists in the past so forcing it would be a bit too restrictive (or did we)

---

We had "_defaultval" for a lot of dists and I don't know how we handled the others.

My thinking about the default so far was this:
1. This PR implements the framework that makes it even possible
2. Next, we implement `get_moment` for a ton of distributions.
3. Next, we set `default_strategy="moment"` and for all distributions that don't have `get_moment`, we put a `kwargs.setdefault("initval", "prior")`, or `initval="prior"` into their signatures.

Then at least all behavior will be defined. Also users who put `initval="moment"` on a dist that doesn't have it _will_ get the `NotImplementedError`.